encoding/test_encoding.py

```diff
@@ -0,0 +1,615 @@
+        listen 443 proto=h2;
+        server ${server_ip}:8000;
+        cache 2;
+
+        tls_certificate ${tempesta_workdir}/tempesta.crt;
+        tls_certificate_key ${tempesta_workdir}/tempesta.key;
+        tls_match_any_server_name;
+        """
+    }
+
+    def setUp(self):
+        # add a chunked body
+        self.backends[0]["response_content"] += self.encode_chunked(BODY_PAYLOAD, CHUNK_SIZE)
+        super().setUp()
+
+    def test(self):
+        self.start_all()
+
+        client = self.get_client("client")
+        server = self.get_server("backend")
+
+        request = [
+            (":authority", "localhost"),
+            (":path", "/"),
+            (":scheme", "https"),
+            (":method", "GET"),
+        ]
+        client.make_request(request)
+        got_response = client.wait_for_response(timeout=5)
+        response = client.responses[-1] if len(client.responses) else None
+
+        self.assertTrue(got_response, "Got no response")
+        self.assertFalse(
+            "Transfer-Encoding" in response.headers,
+            "The response should not have Transfer-Encoding",
+        )
+        cl = response.headers.get("Content-Length", None)
+        self.assertTrue(cl, "The response should have Content-Length")
+        self.assertEqual(
+            response.body, BODY_PAYLOAD, "Dechunked body does not match the original payload"
+        )
+
+        # from now on the response is cached
+        client.make_request(request)
+        got_response = client.wait_for_response(timeout=5)
+
+        self.assertTrue(got_response, "Got no response")
+        self.assertEqual(len(server.requests), 1, "The response has to be served from cache")
+
+
+class TestH2ChunkedIsNotLast(tester.TempestaTest, CommonUtils):
+    """
+    Responses from backend that don't contain Content-Length header
+    and in same time have Transfer-Encoding header with chunked encoding
+    as not the last encoding must not be placed into the cache.
+    Tempesta must add Content-Length to such responses determining length
+    of the body by closing the connection by backend.
+    """
+
+    clients = [
+        {
+            "id": "client",
+            "type": "deproxy_h2",
+            "addr": "${tempesta_ip}",
+            "port": "443",
+            "ssl": True,
+            "ssl_hostname": "localhost",
+        }
+    ]
+    backends = [
+        {
+            "id": "backend",
+            "type": "deproxy",
+            "port": "8000",
+            "response": "static",
+            "response_content": "HTTP/1.1 200 OK\r\n"
+            "Content-type: text/html\r\n"
+            f"Last-Modified: {DATE}\r\n"
+            f"Date: {DATE}\r\n"
+            "Server: Deproxy Server\r\n"
+            "Transfer-Encoding: chunked, gzip\r\n\r\n"
+            "the body does not actually matter",
+        }
+    ]
+    tempesta = {
+        "config": """
+        listen 443 proto=h2;
+        server ${server_ip}:8000;
+        cache 2;
```

Comment:
Need `cache_fulfill * *;`. Also, after last changes such requests can be cached.